### ISSUE TITLE
No CRI-O for Ubuntu 20.10?

### ISSUE THREAD

**Issue by sprior** [cri-o/cri-o#4713]
I was unable to install CRI-O 1.20 on Ubuntu 2-10, I get unable to locate package after adding repo:
deb http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_20.10/ /

**Comment by haircommander** [comment]
I found the issue with 1.20: https://github.com/cri-o/cri-o/pull/4716
what version of cri-o would you like to install? I can attempt to enable it for 1.19 or 1.18 in the meantime

**Comment by sprior** [comment]
I'm doing an initial test install of a cluster and my kubectl and kubeadm versions are at 1.20.5 - do I need to stick with cri-o .1.20 or can I use 1.19 or 1.18?  I'm also willing to blow away the VMs and start from scratch, so the choice is whether I stick with 1.20 and Ubuntu 20.04, or stick with  20.10 and use a slightly backlevel cri-o (if possible).  Any recommendations?

**Comment by haircommander** [comment]
for now I would recommend ubuntu 20.04 and cri-o 1.20.2 :)

**Comment by sprior** [comment]
Ok, I can do that.  Should I leave this issue open so I can see when it's available for 20.10?  Are there other big reasons why k8s isn't ready for 20.10?

**Comment by haircommander** [comment]
no it's a cri-o specific packaging issue, which is my responsibility to fix. we can leave this :)

**Comment by sprior** [comment]
Thanks again, at least I know what to do for now.

**Comment by haircommander** [comment]
happy to help, thanks for trying out CRI-O!

**Comment by sprior** [comment]
Ok, switched to Ubuntu 20.04, following the instructions here: [https://computingforgeeks.com/deploy-kubernetes-cluster-on-ubuntu-with-kubeadm/]

Using the following apt source:
deb http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_20.04/ /

And no cri-o package found:
 apt-cache search cri-
ruby-cri-doc - library for building easy-to-use commandline tools - documentation
cri-o-runc - Open Container Project - runtime
cri-tools - A command line tool used for creating OCI images.

What am I missing?

**Comment by sprior** [comment]
Looks like the instructions at https://computingforgeeks.com/deploy-kubernetes-cluster-on-ubuntu-with-kubeadm are obsolete, I found the install doc at https://github.com/cri-o/cri-o/blob/master/install.md#apt-based-operating-systems and was able to get it.

**Comment by sprior** [comment]
When I tried to run kubeadm init it failed seemingly because it could not automatically detect cri-o as the container runtime.  I found the instructions at https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#configure-cgroup-driver-used-by-kubelet-on-control-plane-node which says I should set cgroupDriver, but I was unable to find a command line option to set this.  I found https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#config-file which quickly turned into a wild goose chase of options I couldn't follow.

So I appear to be REALLY close to getting this to work.  Is there anything closer to a cheat sheet on running kubeadm init in such a way to specify cri-o as the cgroupDriver?

**Comment by sprior** [comment]
Wow, I found https://github.com/cri-o/cri-o/blob/master/tutorials/kubeadm.md and was able to create a config file which has nodeRegistration.criSocket and cgroupDriver: systemd set, so then I tried my kubeadm init command with the --config at the end, but now I'm getting an error:
can not mix '--config' with arguments [control-plane-endpoint pod-network-cidr]
But it is not at all obvious (or apparently documented) what the config file equivalents to these command line parameters are.

**Comment by haircommander** [comment]
oh goodness, sorry you've been chasing around looking for the solutions.

Yes, our kubeadm tutorial is supposed to be the place to learn how to run cri-o with kubeadm. I see it's now out of date... Can you try the startup script:
```
#!/bin/bash                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                              
set -euo pipefail                                                                                                                                                                                                                             
                                                                                                                                                                                                                                              
KUBEADM_CONFIG="${1-/tmp/kubeadm.yaml}"                                                                                                                                                                                                       
echo "Printing to $KUBEADM_CONFIG"                                                                                                                                                                                                            
                                                                                                                                                                                                                                              
if [ -d "$KUBEADM_CONFIG" ]; then                                                                                                                                                                                                             
    echo "$KUBEADM_CONFIG is a directory!"                                                                                                                                                                                                    
    exit 1                                                                                                                                                                                                                                    
fi                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                              
if [ ! -d $(dirname "$KUBEADM_CONFIG") ]; then                                                                                                                                                                                                
    echo "please create directory $(dirname $KUBEADM_CONFIG)"                                                                                                                                                                                 
    exit 1                                                                                                                                                                                                                                    
fi                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                              
if [ ! $(which yq) ]; then                                                                                                                                                                                                                    
    echo "please install yq"                                                                                                                                                                                                                  
    exit 1                                                                                                                                                                                                                                    
fi                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                              
if [ ! $(which kubeadm) ]; then                                                                                                                                                                                                               
    echo "please install kubeadm"                                                                                                                                                                                                             
    exit 1                                                                                                                                                                                                                                    
fi                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                              
kubeadm config print init-defaults --component-configs=KubeletConfiguration > "$KUBEADM_CONFIG"                                                                                                                                               
yq -i eval 'select(.nodeRegistration.criSocket) |= .nodeRegistration.criSocket = "unix:///var/run/crio/crio.sock"' "$KUBEADM_CONFIG"                                                                                                          
yq -i eval 'select(di == 1) |= .cgroupDriver = "systemd"' "$KUBEADM_CONFIG"                                                                                                                                                                   
if [ ! -z "${CIDR-}" ]; then                                                                                                                                                                                                                    
    podSubnet="$CIDR" yq -i eval 'select(di == 1) |= .networking.podSubnet = env(podSubnet)' "$KUBEADM_CONFIG"                                                                                                                                
fi
```

with CIDR set to your pod network cidr? I don't know about control-plane-endpoint, and I'm not actually sure podSubnet==pod-network-cidr, but I am hopeful

**Comment by sprior** [comment]
I'm new at this, but when running kubeadm init I had set --pod-network-cidr=192.168.2.0/16 which is an available but unused network I thought would be used for the "internal" network between pods, but I could be totally wrong about that.  I also set --control-plane-endpoint=k8s-cluster.geekster.com  which is a CNAME for the address of the master node in the cluster.  I thought both of these parameters were necessary, but again I'm new at this.
